src/sentry/tagstore/multi/backend.py

```diff
@@ -0,0 +1,238 @@
+"""
+sentry.tagstore.multi.backend
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:copyright: (c) 2010-2017 by the Sentry Team, see AUTHORS for more details.
+:license: BSD, see LICENSE for more details.
+"""
+
+from __future__ import absolute_import
+
+import six
+import random
+from threading import Thread
+from six.moves.queue import Queue, Full
+from django.conf import settings
+from operator import itemgetter
+
+from sentry.tagstore.base import TagStorage
+from sentry.utils.imports import import_string
+
+
+class QueuedRunner(object):
+    """\
+    Secondary backend runner that puts method calls on a bounded queue and drops them
+    when the queue is full.
+
+    A separate (non-main) thread works the queue.
+    """
+
+    def __init__(self):
+        self.q = q = Queue(maxsize=100)
+        self.sample_channel = getattr(settings, 'SENTRY_TAGSTORE_MULTI_SAMPLING', 1.0)
+
+        def worker():
+            while True:
+                (func, args, kwargs) = q.get()
+                try:
+                    func(*args, **kwargs)
+                except Exception:
+                    pass
+                finally:
+                    q.task_done()
+
+        t = Thread(target=worker)
+        t.setDaemon(True)
+        t.start()
+
+    def run(self, f, *args, **kwargs):
+        if random.random() <= self.sample_channel:
+            try:
+                self.q.put((f, args, kwargs), block=False)
+            except Full:
+                return
+
+
+class ImmediateRunner(object):
+    """\
+    Secondary backend runner that runs functions immediately. Useful for tests.
+    """
+
+    def run(self, f, *args, **kwargs):
+        return f(*args, **kwargs)
+
+
+class MultiTagStorage(TagStorage):
+    """
+    A backend which will write to multiple backends, and read from the first (by default).
+    Writes to non-primary backends will happen on a background thread. This is designed to
+    allow you to dual-write for the purpose of migration.
+
+    >>> MultiTagStorage(backends=[
+    >>>     ('sentry.tagstore.legacy.LegacyTagStorage', {}),
+    >>>     ('sentry.tagstore.v2.V2TagStorage', {}),
+    >>> ])
+    """
+
+    def __init__(self, backends, read_selector=itemgetter(0),
+                 runner='QueuedRunner', **kwargs):
+        assert backends, "you should provide at least one backend"
+
+        self.backends = []
+        for backend, backend_options in backends:
+            if isinstance(backend, six.string_types):
+                backend = import_string(backend)
+            self.backends.append(backend(**backend_options))
+
+        self.read_selector = read_selector
+        self.runner = globals()[runner]()
```

Comment:
I'd rather we not just pluck this out of globals. We can easily just maintain like:

```python
self.runner = {'QueuedRunner': QueuedRunner, 'ImmediateRunner': ImmediateRunner}[runner]()
```

Just value being some sanity checking and not dipping into whatever variables happen to exist in scope.